air/src/proof/mod.rs

```diff
@@ -207,61 +207,126 @@ fn get_conjectured_security(
+    // This considers only the first term given in eq. 7 in https://eprint.iacr.org/2022/1216.pdf,
+    // i.e. 0.5 * (m + 0.5)^7 * n^2 / (rho^1.5.q) as all other terms are negligible in comparison.
+    let fri_commit_err_bits = extension_field_bits
+        - ((0.5 * (m + 0.5).powf(7.0) / rho.powf(1.5)) * lde_domain_size.powf(2.0)).log2();
+
+    // Compute FRI query-phase soundness error
+    let fri_queries_err_bits =
+        options.grinding_factor() as f64 - ((1.0 - theta).powf(num_fri_queries)).log2();
+
+    // Combined error for FRI
+    let fri_err_bits = cmp::min(fri_commit_err_bits as u64, fri_queries_err_bits as u64);
+    if fri_err_bits < 1 {
+        return 0;
+    }
+    let fri_err_bits = fri_err_bits - 1;
+
+    // To apply Theorem 8 in https://eprint.iacr.org/2022/1216.pdf, we need to apply FRI with
+    // a slightly larger agreement parameter alpha.
+    // More concretely, we need alpha > rho_plus.sqrt() where rho_plus is the rate in function field
+    // F(Z) and defined as (trace_domain_size + 2.0) / lde_domain_size .
+    // This means that the range of m needs to be restricted in order to insure that
+    // alpha := 1 - theta := rho.sqrt() * (1 + 1/2m) is greater than rho_plus.sqrt().
+    // Determining the range of m is the responsibility of the calling function.
+    // Now, once m is fixed, we need to make sure that we choose an m_plus such that
+    // alpha <= rho_plus.sqrt() * (1 + 1/2m_plus). This m_plus will be used to define
+    // the list-decoding list size in F(Z).
+
+    // Modified rate in function field F(Z)
+    let rho_plus = (trace_domain_size + 2.0) / lde_domain_size;
+    // New proximity parameter m_plus, corresponding to rho_plus, needed to make sure that
+    //  alpha < rho_plus.sqrt() * (1 + 1 / (2 * m_plus))
+    let m_plus = (1.0 / (2.0 * (alpha / rho_plus.sqrt() - 1.0))).ceil();
+
+    // List size
+    let l_plus = (2.0 * m_plus + 1.0) / (2.0 * rho_plus.sqrt());
+
+    // ALI related soundness error. Note that C here is equal to 1 because of our use of
+    // linear batching.
+    let ali_err_bits = -l_plus.log2() + extension_field_bits;
+
+    // DEEP related soundness error. Note that this uses that the denominator |F| - |D ∪ H|
+    // can be approximated by |F| for all practical domain sizes.
+    let deep_err_bits =
+        -(l_plus * (max_deg * (trace_domain_size + 1.0) + (trace_domain_size - 1.0))).log2()
+            + extension_field_bits;
+
+    let min = cmp::min(
+        cmp::min(fri_err_bits, ali_err_bits as u64),
+        deep_err_bits as u64,
+    );
+    if min < 1 {
+        return 0;
+    }
 
-    query_security += options.grinding_factor();
+    min - 1
+}
 
-    cmp::min(
-        cmp::min(pre_query_security, query_security) - 1,
-        collision_resistance,
-    )
+// Computes the largest proximity parameter m needed for Theorem 8
+// in https://eprint.iacr.org/2022/1216.pdf to work.
+fn upper_m(h: u64) -> f64 {
```

Comment:
A couple of small nits:
- Let's use `///` for doc comments.
- For proper link formatting we should put links in `<>`. Here it would be `/// in <https://eprint.iacr.org/2022/1216.pdf> to work.`
- I would probably name the function `compute_upper_m()`.
- Let's add `#[cfg(feature = "std")]` attribute since we currently call this only in the `std` context.

Also, maybe add to the comment that for practical trace lengths (e.g., up to $2^{30}$) the returned value will be smaller than about 2300.

Reply:
Sounds good!
I am not sure about the upper bound you mentioned, I think that for traces of length $2^{22}$ an upper bound is around 2 million. It might make sense to put a cap on the return value as usually the optimal $m$ is in the lower range of `[3,..., upper_m(h)]`.